### ISSUE TITLE
AssertionError (must be unit) caused by strange combination of _while_loop, jit and nested init_state

### ISSUE THREAD

**Issue by jonasrauber** [google/jax#331]
The following minimal example causes a strange `AssertionError`:

```python3
import jax
import jax.numpy as np


def cond_fun(state):
    return jax.lax.lt(state[1][0], 10.)


def body_fun(state):
    return (state[0], (state[1][0] + 1.,))


@jax.jit
def run(x):
    c = 1.
    init_state = (x, (c,))
    return jax.lax._while_loop(cond_fun, body_fun, init_state)


run(np.array([1, 2, 3]))
```

```
Traceback (most recent call last):
  File "main.py", line 20, in <module>
    run(np.array([1, 2, 3]))
  File "jax/jax/api.py", line 80, in f_jitted
    jaxtupletree_out = xla.xla_call(jaxtree_fun, *jaxtupletree_args)
  File "jax/jax/core.py", line 529, in call_bind
    ans = primitive.impl(f, *args, **kwargs)
  File "jax/jax/interpreters/xla.py", line 417, in xla_call_impl
    compiled_fun = xla_callable(fun, *map(abstractify, flat_args))
  File "jax/jax/linear_util.py", line 146, in memoized_fun
    ans = call(f, *args)
  File "jax/jax/interpreters/xla.py", line 431, in xla_callable
    compiled, result_shape = compile_jaxpr(jaxpr, consts, *abstract_args)
  File "jax/jax/interpreters/xla.py", line 118, in compile_jaxpr
    built_c = jaxpr_computation(jaxpr, const_vals, (), *arg_shapes)
  File "jax/jax/interpreters/xla.py", line 135, in jaxpr_computation
    map(write, jaxpr.constvars, map(c.Constant, const_vals))
  File "jax/jax/util.py", line 43, in safe_map
    return list(map(f, *args))
  File "jax/jax/lib/xla_bridge.py", line 329, in Constant
    return _constant_handlers[py_type](self, py_val, canonicalize_types)
  File "jax/jax/interpreters/xla.py", line 160, in unit_constant
    assert not val  # must be unit
AssertionError
```

The interesting thing is: it only happens for this particular combination:

- run must be jitted
- init_state must contain a traced variable (`x`) and a non-traced variable (`c`)
- the non-traced variable must be nested
    - using `(x, c)` and adapting `cond_fun` and `body_fun` accordingly works fine

It's not really a big deal because I can just use a flat state variable for now, but it seems like an extremely strange `AssertionError` and it took some time to nail it down to this example.

**Comment by dougalm** [comment]
Thanks for the bug report! Should be fixed by #334

**Comment by jonasrauber** [comment]
Thanks!
